added static get_max_length for mp time series

vincentberenz opened the pull request:
[//]: # "Thanks for your contribution.  To make life of the reviewers easier,"
[//]: # "please give this pull request a meaningful title and provide the"
[//]: # "requested information below."

## Description

A leader and a follower multiprocess time series sharing the same segment_id must "agree" on their size upon construction.
This is error prone. A static get_max_length function has been added, which returns the size used by the leader during its construction, which can be used by follower to setup their own size;


[//]: # "Description of what you did.  If it is more complex, consider to add a"
[//]: # "'Summary' section."


## How I Tested

Updated the demo and the unit tests

[//]: # "Explain how you tested your changes"


## Do not merge before

[//]: # "Link other open pull request here that need to be merged first."
[//]: # "Remove this section if it does not apply."

- [ ] link to dependent pull request

## I fulfilled the following requirements

[//]: # "Please make sure you followed these steps before requesting a review."
[//]: # "Check the boxes in the list below, when done."

- [X ] All new code is formatted according to our style guide (for C++ run clang-format, for Python, run flake8 and fix all warnings).
- [X ] All new functions/classes are documented and existing documentation is updated according to changes.
- [ X] No commented code from testing/debugging is kept (unless there is a good reason to keep it).

Commit:
added static get_max_length for mp time series

luator reviewed (approved):
Looks good, thanks a lot!

vincentberenz commented:
@luator 

Maybe we could also add a new constructor which does not take a max_length argument (directly reads it from the shared memory) ?

luator commented:
Yes, this would make a lot of sense.  I'm a bit unsure how to best handle the two cases to avoid potential confusion.  Is there any use-case where it would be needed to explicitly specify the length for a non-leader instance?  If not it would maybe be best to make constructor protected and provide static factory methods instead.  I guess the same as for `max_length` could also be done for `start_timeindex`?

    createLeader(std::string segment_id,
                 size_t max_length,
                 Index start_timeindex = 0)

    createNonLeader(std::string segment_id)

("NonLeader" is not a great name, I just don't have a better idea right now)

vincentberenz commented:
@luator 

A factory and doing the same for the start_index sound like a good idea.
I can do this next week (then pushing on this branch ?), except if you'd prefer to do it ? (then we can already merge this branch and close this pull request)

luator commented:
It would be great if you could do this.  Adding it it to this branch sounds good.

Commit:
added factories for multiprocess time series, regular constructor marked deprecated

Commit:
added doc to factories

vincentberenz commented:
@luator 

I added the factories, and updated the unit tests and demo accordingly.
For the moment I just set the constructor as deprecated, to avoid changing the API.

luator commented on include/time_series/multiprocess_time_series.hpp:
```diff
@@ -87,20 +91,71 @@ class MultiprocessTimeSeries
+                                                   Index start_timeindex = 0)
+    {
+        bool leader = true;
+        return MultiprocessTimeSeries<T>(
```
Isn't this using the constructor which is now marked deprecated?

luator replied:
If it is the case, we would always get the warning, even when using it correctly.

vincentberenz replied:
Indeed, things are not ideal now. Idea that once ready, we make the constructor protected, and remove the deprecated tag. I marked it deprecated mostly so that we do not forget to make it protected later on (could have made an issue instead)

luator replied:
alright :+1:

luator reviewed (approved):
Looks good, thanks a lot :)

Just one concern regarding the deprecation, but maybe I'm misinterpreting things.
